src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -696,89 +640,139 @@ HeapTupleOfForeignConstraintIncludesColumn(HeapTuple heapTuple, Oid relationId,
-				relationId);
-	SysScanDesc scanDescriptor = systable_beginscan(pgConstraint, scanIndexId, useIndex,
-													NULL,
-													scanKeyCount, scanKey);
+/*
+ * ConstraintIsAForeignKey returns true if the given constraint name
+ * is a foreign key defined on the relation.
+ */
+bool
+ConstraintIsAForeignKey(char *inputConstaintName, Oid relationOid)
+{
+	bool excludeSelfReference = false;
+	bool findAll = true;
+	List *foreignKeyOids = GetTableForeignConstraintOidsInternal(relationOid,
+																 Anum_pg_constraint_conrelid,
+																 excludeSelfReference,
+																 findAll);
 
-	HeapTuple heapTuple = systable_getnext(scanDescriptor);
-	while (HeapTupleIsValid(heapTuple))
+	Oid foreignKeyOid = FindForeignKeyOidWithName(foreignKeyOids, inputConstaintName);
+
+	return OidIsValid(foreignKeyOid);
+}
+
+
+/*
+ * FindForeignKeyOidWithName searches the foreign key constraint with
+ * inputConstraintName in the given list of foreign key constraint OIDs.
+ * Returns the OID of the matching constraint. If there no matching constraint
+ * in the given list, then returns InvalidOid.
+ */
+static Oid
+FindForeignKeyOidWithName(List *foreignKeyOids, const char *inputConstraintName)
+{
+	Oid foreignKeyOid = InvalidOid;
+	foreach_oid(foreignKeyOid, foreignKeyOids)
 	{
-		Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(heapTuple);
+		char *constraintName = get_constraint_name(foreignKeyOid);
 
-		if (constraintForm->contype == CONSTRAINT_FOREIGN)
+		if (strncmp(constraintName, inputConstraintName, NAMEDATALEN) == 0)
 		{
-			systable_endscan(scanDescriptor);
-			heap_close(pgConstraint, NoLock);
-
-			return true;
+			return foreignKeyOid;
 		}
-
-		heapTuple = systable_getnext(scanDescriptor);
 	}
 
-	systable_endscan(scanDescriptor);
-	heap_close(pgConstraint, NoLock);
-
-	return false;
+	return InvalidOid;
 }
 
 
 /*
- * ConstraintIsAForeignKey returns true if the given constraint name
- * is a foreign key to defined on the relation.
+ * GetTableForeignConstraintOidsInternal takes in a relationOid, and returns
+ * the list of OIDs for foreign constraints that the relation with relationOid
+ * is involved according to "attributeNumber" argument. In that sense,
+ * "attributeNumber" should be:
+ *  - "Anum_pg_constraint_conrelid" to get the ones in which our relation is
+ *    the "referencing" one, and
+ *  - "Anum_pg_constraint_confrelid" to get the ones in which our relation is
+ *    the "referenced" one.
+ * Note that since there is no index for the column Anum_pg_constraint_confrelid,
+ * this function performs sequential scan on pg_constraint in that case.
+ *
+ * Self-referencing foreign keys will not be considered if excludeSelfReference
+ * is true.
  */
-bool
-ConstraintIsAForeignKey(char *constraintNameInput, Oid relationId)
+static List *
+GetTableForeignConstraintOidsInternal(Oid relationOid, AttrNumber attributeNumber,
```

Comment:
As far as I see the only benefit of having `findAll` parameter is performance. But these codepaths are not going to be performance critical. And this findAll parameter is making it complex. What about we remove it?